doc/ensemble.rst

```diff
@@ -0,0 +1,739 @@
+the orchestrator is colocated or not, each script added to the ensemble is
+loaded into the orchestrator prior to the execution of any ensemble member.
+The flexibility of device selection further enhances
+adaptability, offering the choice between "GPU" or "CPU." Users have the flexibility to choose
+between `"GPU"` or `"CPU"` for device selection, and in environments with multiple
+devices, specific device numbers can be specified via `devices_per_node` such as `"GPU:1,"`.
+
+When specifying a TorchScript using ``Ensemble.add_script()``, the
+following arguments are offered:
+
+-  `name`  (str) : key to store script under
+-  `script` (str, optional) : TorchScript code
+-  `script_path` (str, optional) : file path to TorchScript code
+-  `device`  (str, optional) : device for script execution, defaults to “CPU”
+-  `devices_per_node` (int) : assign the number of CPU's or GPU's to use on the node
+
+You might use TorchScript scripts to represent individual models within the ensemble:
+
+Example: Loading an TorchScript to the Ensemble
+-----------------------------------------------
+In this example, we demonstrate how to instruct SmartSim to load
+an TorchScript into the database at ensemble runtime.
+
+**Python Script: Define a TorchScript for Ensemble Purposes**
+
+Define the TorchScript code to a variable in the Python driver script:
+
+.. code-block:: python
+
+    torch_script_str = "def negate(x):\n\treturn torch.neg(x)\n"
+
+**SmartSim Ensemble Integration:**
+
+Assuming an initialized ``Ensemble`` named `smartsim_ensemble`, we specify the
+following parameters to the ``Ensemble.add_script()`` function:
+
+.. code-block:: python
+
+    smartsim_ensemble.add_script(name="example_script", script=torch_script_str, device="CPU")
+
+In this integration, we provide the following details:
+
+-  `name`: "example_script" - key to store script under
+-  `script`: torch_script_str - TorchScript code
+-  `device`: "CPU" - device for script execution
+
+When the ensemble is started via ``Experiment.start()``, the TorchScript will be loaded to the
+orchestrator that is launched prior to the start of the ensemble.
+
+=========================
+Data Collision Prevention
+=========================
+.. _prefix_ensemble:
+--------
+Overview
+--------
+When multiple ensemble members use the same code to access their respective models
+in the Orchestrator, key overlapping can occur, leading to inadvertent data access
+between ensemble members. To address this, the SmartSim Ensembles supports key prefixing
+via the ``Ensemble.enable_key_prefixing()`` function,
+which automatically adds the model name as a prefix to the keys used for access.
+Enabling key prefixing eliminates issues related to key overlapping, allowing ensemble
+members to use the same code without issue.
+
+-------------------------------
+Example: Ensemble Key Prefixing
+-------------------------------
+In this example, we explore ensemble key prefixing in SmartSim.
```

Comment:
This example explores...

Reply:
note: the opening sentence is redundant and offers no meat. Of course it explores key prefixing in smartsim, it's in the key prefixing section and has a title "Example: Ensemble Key Prefixing".... Can we give a better intro? "... demonstrates key collisions and their mitigation throuch key-prefixing" etc

Reply:
note 2: it would be nice to have a demo that suffers a collision, followed by a mitigation. 

note 3: we had a very similar section on experiments to describe `enable_key_prefixing` - since they use the same internal behaviors, could we unify key prefixing docs somehow?